diff --git a/src/capnp/rpc.c++ b/src/capnp/rpc.c++
index b50bca6..136c20d 100644
--- a/src/capnp/rpc.c++
+++ b/src/capnp/rpc.c++
@@ -1273,6 +1273,7 @@ private:
                    kj::Own<QuestionRef>&& questionRef,
                    kj::Array<PipelineOp>&& ops)
         : RpcClient(connectionState), questionRef(kj::mv(questionRef)), ops(kj::mv(ops)) {}
+    ~PipelineClient() noexcept(false) {}
 
     WriteDescriptorResult writeDescriptor(rpc::CapDescriptor::Builder descriptor,
                                           OutgoingFds& fds) override {
